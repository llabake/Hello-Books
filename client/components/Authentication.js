import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import toastMessage from '../helpers/toastMessage';

/**
 * 
 * 
 * @export
 * @param {any} Component to be rendere
 * @returns 
 */
export default function requireAuthentication(Component) {
  /**
   * 
   * 
   * @class Authenticate
   * @extends {Component}
   */
  class Authenticate extends Component {
    static propTypes = {
      authenticated: PropTypes.bool.isRequired,
      dispatch: PropTypes.func.isRequired
    };
  /**
   * 
   * 
   * @returns { js } protected component 
   * @memberof Authenticate
   */
  render() {
      return (
        <div>
          {this.props.authenticated
            ? <Component {...this.props} />
            :
            <Redirect to={{
              pathname: '/signin',
              state: { from: this.props.location }
            }}
            />
          }
          {/* toastMessage(' Please login to access this route', 'failure') */}
          
        </div>
      );
    }
  }

  const mapStateToProps = ({ userReducer }) => ({
    authenticated: userReducer.authenticated,
  });

  return connect(mapStateToProps)(Authenticate);
}
