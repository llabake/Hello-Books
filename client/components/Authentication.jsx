import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import toastMessage from '../helpers/toastMessage';
import { authenticateUser } from '../helpers/utils';

/**
 * 
 * 
 * @export
 * 
 * @param {any} ComposedComponent the child component to be rendered
 * 
 * @returns {void}
 */
export default function requireAuthentication(ComposedComponent) {
  /**
   * 
   * 
   * @class Authenticate
   * 
   * @extends {Component}
   */
  class Authenticate extends Component {
    static propTypes = {
      // authenticated: PropTypes.bool.isRequired,
      dispatch: PropTypes.func.isRequired
    };
    /**
     * 
     * 
     * @param {any} nextProps
     * 
     * @memberof Authenticate
     * 
     * @returns {void}
     */
    componentWillReceiveProps(nextProps) {
      if (!nextProps.isAuthenticated) {
        <Redirect to={{
          pathname: '/signin',
          state: { from: this.props.location }
        }}
        />
      }
    }
    /**
     * 
     * 
     * @returns { object } returns the protected object
     * 
     * @memberof Authenticate
     */
    render() {
      const { isAuthenticated } = authenticateUser()
      return (
        <div>
          {isAuthenticated
            ? <ComposedComponent {...this.props} />
            :
            <Redirect to={{
              pathname: '/signin',
              state: { from: this.props.location }
            }}
            />
          }
          {/* toastMessage(' Please login to access this route', 'failure') */}

        </div>
      );
    }
  }


  return connect(mapStateToProps, null)(Authenticate);
}

export const mapStateToProps = (state) => {
  return {
    error: state.userReducer.error,
    user: state.userReducer.authUser,
    authenticated: state.userReducer.authenticated
  };
};