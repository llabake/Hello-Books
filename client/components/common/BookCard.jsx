import React from 'react';
import { Link } from 'react-router-dom';
import { bookDefaultImage } from '../../helpers/utils';

import leanstart from '../../media/lean start.jpg';

const BookCard = (props) => {
  const { book } = props;
  return (
    <div className="book col s12 m3 l3">
      <div className="card" style={{ minHeight: '250px', maxHeight: '250px' }}>
        <div className="card-image waves-effect waves-block waves-light" style={{ maxHeight: '175px', minHeight: '170px', paddingTop: '10px' }}>
          <Link to={'/book/' + book.id}>
            <img src={book.image || bookDefaultImage}
              style={{ maxHeight: '170px', minHeight: '165px' }}
              alt="Book Image"
              title={book.title} 
            />
          </Link>
        </div>
        <div className="card-content"
          style={{ position: 'absolute', bottom: 0, width: '100%' }}>
          <p><Link to={'/book/' + book.id} className="btn">Details</Link></p>
        </div>
      </div>
      <div className="book-info center-align">
        <div className="book-title"><span>{book.title} </span></div>
        <div className="book-author">by <span>{book.author}</span></div>
      </div>
    </div>

  );
}

export default BookCard;