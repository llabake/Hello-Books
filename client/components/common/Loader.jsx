import React from 'react';

import ajaxLoader from '../../media/ajax-loader.gif';

const Loader = () => {
  return (
    <div
      className="center-align"
      style={{ marginTop: '1em', marginBottom: '1em' }}
    >
      <img src={ajaxLoader} alt="Loading..." />
    </div>
  )
}

export default Loader;