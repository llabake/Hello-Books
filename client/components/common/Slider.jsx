import React, { Component } from 'react';



import sliderimage1 from '../../media/archive-beautiful-book-stack-256455.jpg';
import sliderimage2 from '../../media/bookcase-books-bookshelves-256541.jpg';
import sliderimage3 from '../../media/john-michael-thomson-555276-unsplash.jpg';
import sliderimage4 from '../../media/pj-accetturo-172481-unsplash.jpg';
import sliderimage5 from '../../media/joshua-coleman-623144-unsplash.jpg';
import sliderimage6 from '../../media/radu-marcusu-498248-unsplash.jpg';


/**
 *
 *
 * @export
 * @class Slider
 * @extends {Component}
 */
export default class Slider extends Component {
  
  /**
   * 
   * 
   * @memberof Slider
   * @returns {function} initialises the slider 
   */
  componentDidUpdate() {
    $('.slider').slider({
      full_width: true
    })
  }
  
  
  /**
   * 
   * 
   * @returns {Object} html
   * @memberof Slider 
   */
  render () {
    return (
      <div className="home-page-slider">
        <div className="slider" style={{background: 'rgba(0,0,0,0.7)'}}>
          <ul className="slides">
            <li>
              <img src={sliderimage1} alt="sliderimage1" style={{ backgroundColor: 'rgba(0,0,0,3)'}}/>
              <div className="caption left-align">
                <h1 style={{'textShadow':'0 1px 0 black',}}><strong>Welcome to HelloBooks</strong></h1>
                <h3 style={{'textShadow':'0 1px 0 black'}} className="light grey-text text-lighten-3"><strong>Sign Up today to enter into <br /> a whole book world</strong></h3>
              </div>
            </li>
            <li>
              <img src={sliderimage2} alt=""style={{ backgroundColor: 'rgba(0,0,0,3)'}} />
              <div className="caption left-align">
                <h1 style={{ 'textShadow': '0 1px 0 black', 'color': '#ffffff'}} ><strong>Welcome to the HelloBooks</strong></h1>
                <h3 className="light grey-text text-lighten-3"><strong>You can borrow a book to read <br /> and post reviews on it</strong></h3>
              </div>
            </li>
            <li>
              <img src={sliderimage3} alt=""style={{ backgroundColor: 'rgba(0,0,0,3)'}} />
              <div className="caption left-align">
                <h1 style={{ 'textShadow': '0 1px 0 black', 'color': '#ffffff'}}><strong>Welcome to HelloBooks</strong></h1>
                <h3 className="light grey-text text-lighten-3"><strong>Create Reading list and <br />  suggest books to the library</strong></h3>
              </div>
            </li>
            <li>
              <img src={sliderimage4} alt="" style={{ backgroundColor: 'rgba(0,0,0,3)'}}/>
              <div className="caption left-align">
                <h1 style={{ 'textShadow': '0 1px 0 black', 'color': '#ffffff'}}><strong>Welcome to HelloBooks</strong></h1>
                <h3 className="light grey-text text-lighten-3"><strong>Review books and add <br />books to your favorite list</strong></h3>
              </div>
            </li>
            <li>
              <img src={sliderimage5} alt="" style={{ backgroundColor: 'rgba(0,0,0,3)'}}/>
              <div className="caption left-align">
                <h1 style={{ 'textShadow': '0 1px 0 black', 'color': '#ffffff'}}><strong>Welcome to HelloBooks</strong></h1>
                <h3 className="light grey-text text-lighten-3"><strong>Sign Up today to enter into a whole book world</strong></h3>
              </div>
            </li>
            <li>
              <img src={sliderimage6} alt="" style={{ backgroundColor: 'rgba(0,0,0,3)'}}/>
              <div className="caption left-align">
                <h1 style={{ 'textShadow': '0 1px 0 black', 'color': '#ffffff'}}><strong>Welcome to HelloBooks</strong></h1>
                <h3 className="light grey-text text-lighten-3"><strong>Sign Up today to enter into a whole book world</strong></h3>
              </div>
            </li>

          </ul>
        </div>
      </div>
    )
  }

}