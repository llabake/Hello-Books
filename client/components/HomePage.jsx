import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux'

import Slider from '../components/common/Slider';
import Carousel from '../components/common/Carousel';
import BookCard from '../components/common/BookCard';
import { fetchPopularBooks } from '../actions/bookAction';

import leanstart from '../media/lean start.jpg';
import ajaxLoader from '../media/ajax-loader.gif';


/**
 * 
 * 
 * @export
 * @class HomePage
 * @extends {Component}
 */
class HomePage extends Component {
  // static propTypes = {
  //   ,
  // }
  /**
   * Creates an instance of HomePage.
   * @param {any} props
   * 
   * @memberof HomePage
   */
  constructor(props) {
    super(props);
    this.state = {};
  }


  /**
   * @returns {Array} most popular books array
   * 
   * @memberof HomePage
   */
  componentWillMount() {
    this.props.fetchPopularBooks();
  }

  /**
   * 
   * @returns {function} initialises the side bav
   * 
   * @memberof HomePage
   */
  // componentDidMount() {
  //   $(".button-collapse").sideNav();
  // }




  /**
  * 
  * 
  * @returns {Object} html
  * 
  * @memberof HomePage
  */
  render() {
    const { popularBooks, loadingPopularBooks } = this.props;
    return (
      <div>
        <Slider />
        <div >
          <h4 className="center-align">
            <span>Most Popular BOOKS</span>
          </h4>
        </div>
        <div className="divider" />
        {loadingPopularBooks ? 
          <div className="center-align" style={{ marginTop: '1em', marginBottom: '1em' }}>
            <img src={ajaxLoader} alt="Loading..." />
          </div> : ''}
        <div className="section">
          <div className="container">
            <div className="row">
              {!loadingPopularBooks ? popularBooks.length ?
                popularBooks.map((book) => {
                  return <div key={book.id}><BookCard book={book} /></div>
                })
                :
                <h5 className="center-align"> No books found </h5>
                : ''}
            </div>
          </div>
        </div>
        <div className="divider" />
        <div className="section similar">
          <h3 className="center-align">
            Based on readers favorites
          </h3>
          <div><Carousel /></div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    popularBooks: state.bookReducer.popularBooks,
    loadingPopularBooks: state.bookReducer.loadingPopularBooks
  };
};


const mapDispatchToProps = (dispatch) => {
  return {
    fetchPopularBooks: () => dispatch(fetchPopularBooks()),
  };
};


HomePage.propTpyes = {
  fetchPopularBooks: PropTypes.func,
  popularBooks: PropTypes.arrayOf(PropTypes.object)
}
export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
