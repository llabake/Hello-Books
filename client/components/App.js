import React, { Component } from 'react';
import { render } from 'react-dom';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import requireAuthentication from './Authentication';
import HomePage from './HomePage';
import SignUpForm from './SignUpForm';
import SignInForm from './SignInForm';
import AllBooks from './main/AllBooks'
import AddBook from './adminDashBoard/AddBook';
import SingleBook from './main/SingleBook';
import AdminDashBoard from './adminDashBoard/AdminDashBoard';
import UserProfilePage from './userDashBoard/UserProfilePage';
import Favorite from './userFavorite/Favorite';
import SearchResult from './common/SearchResult';
import EditProfile from './userDashBoard/EditProfile';
import NotFound from './NotFound';

import '../styles/style.scss';
import Header from './common/Header';
import Footer from './common/Footer';
/**
 * 
 * 
 * @export
 * @class App
 * @extends {Component}
 */
export default class App extends Component {

  componentDidMount() {
    $('.button-collapse').sideNav({
      menuWidth: 250, // Default is 300
      edge: 'left', // Choose the horizontal origin
      draggable: true // Choose whether you can drag to open on touch screens
    }
  );
  // $('.carousel').carousel({
  //   dist: 0, // Perspective zoom. 
  //   shift: 0, // Set the spacing of the center item.
  //   padding: 50, // Set the padding between non center items.
  //   indicators: true, // Show indicators. Default is false
  // });
  // autoplay()
  // function autoplay() {
    // $('.carousel').carousel('next');
    // setTimeout(autoplay, 4500)}

    $('.slider').slider();
    $('.dropdown-button').dropdown({
      belowOrigin: true,
    });
    $('.tooltipped').tooltip({ delay: 50 });
    $('.tabs').tabs();
    $('.tooltipped').tooltip();
  

    // $('.modal').modal();
    // $('select').material_select();
    
  }
  /**
   * 
   * 
   * @returns {object} main App
   * @memberof App
   */
  render () {
    return (
      <Router>
        <div className="body">
        <header>
        <Header />
        </header>
        <main>
        <Switch>
          <Route exact path='/' component={HomePage} />
          <Route exact path='/signup' component={SignUpForm} />
          <Route exact path='/signin' component={SignInForm} />
          <Route exact path='/allbooks' component={requireAuthentication(AllBooks)} />
          <Route exact path='/addbook' component={requireAuthentication(AddBook)} />
          <Route exact path='/admindashboard' component={requireAuthentication(AdminDashBoard)} />
          <Route exact path='/book/:bookId' component={requireAuthentication(SingleBook)} />
          <Route exact path='/profile' component={requireAuthentication(UserProfilePage)} />
          <Route exact path='/favorite' component={requireAuthentication(Favorite)}/>
          <Route exact path='/search/:searchTerm' component={requireAuthentication(SearchResult)}/>
          <Route exact path='/editprofile' component={requireAuthentication(EditProfile)} />
          <Route component={NotFound} />
        </Switch>
        </main>
        <footer>
        <Footer />
        </footer>
        </div>
      </Router>
    );
  }
}





