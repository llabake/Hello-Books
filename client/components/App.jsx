import React, { Component } from 'react';
import { render } from 'react-dom';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import requireAuthentication from './Authentication';
import HomePage from './HomePage';
import SignUpForm from './SignUpForm';
import SignInForm from './SignInForm';
import AllBooks from './main/AllBooks'
import AddBook from './adminDashBoard/AddBook';
import SingleBook from './main/SingleBook';
import AdminDashBoard from './adminDashBoard/AdminDashBoard';
import UserProfilePage from './userDashBoard/UserProfilePage';
import Favorite from './userFavorite/Favorite';
import SearchResult from './common/SearchResult';
import EditProfile from './userDashBoard/EditProfile';
import NotFound from './NotFound';

import '../styles/style.scss';
import Header from './common/Header';
import Footer from './common/Footer';
/**
 * 
 * 
 * @export
 * @class App
 * @extends {Component}
 */
export default class App extends Component {
  /**
   *@returns {void}
   *
   * @memberof App
   */
  componentDidMount() {
    $('.button-collapse').sideNav({
      menuWidth: 250, // Default is 300
      edge: 'left', // Choose the horizontal origin
      draggable: true // Choose whether you can drag to open on touch screens
    }
    );
    $('.slider').slider({
      full_width: true,
    }
    );
    $('.dropdown-button').dropdown({
      belowOrigin: true,
    });
    $('.tooltipped').tooltip({ delay: 50 });
    $('ul.tabs').tabs();
    $('.tooltipped').tooltip({ delay: 50 });
  }

  /**
   * 
   * 
   * @returns {object} main App
   * @memberof App
   */
  render() {
    return (
      <Router>
        <div className="body">
          <header>
            <Header />
          </header>
          <main>
            <Switch>
              <Route exact path='/' component={HomePage} />
              <Route exact path='/signup' component={SignUpForm} />
              <Route exact path='/signin' component={SignInForm} />
              <Route exact path='/allbooks' component={requireAuthentication(AllBooks)} />
              <Route exact path='/addbook' component={requireAuthentication(AddBook)} />
              <Route exact path='/admindashboard' component={requireAuthentication(AdminDashBoard)} />
              <Route exact path='/book/:bookId' component={requireAuthentication(SingleBook)} />
              <Route exact path='/profile' component={requireAuthentication(UserProfilePage)} />
              <Route exact path='/favorite' component={requireAuthentication(Favorite)} />
              <Route exact path='/search/:searchTerm' component={requireAuthentication(SearchResult)} />
              <Route exact path='/editprofile' component={requireAuthentication(EditProfile)} />
              <Route component={NotFound} />
            </Switch>
          </main>
          <footer>
            <Footer />
          </footer>
        </div>
      </Router>
    );
  }
}





