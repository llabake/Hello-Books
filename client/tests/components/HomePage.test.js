import React from 'react';

import { HomePage, mapDispatchToProps, mapStateToProps } from '../../components/HomePage';
import reducer from '../../reducers'
import Loader from '../../components/common/Loader';
import Slider from '../../components/common/Slider';
import BookCard from '../../components/common/BookCard';

test('The HomePage component mounts correctly if all props it needs are passed into it.', () => {
  const props = {
    authenticated: false,
    user: {},
    errors: {},
    fetchPopularBooks: jest.fn(),
    popularBooks: [],
    loadingPopularBooks: false
  }

  const wrapper = shallow(<HomePage {...props} />)
  expect(wrapper).toBeDefined();
})

test('Homepage should call componentWillMount', () => {
  let spy;
  afterEach(() => {
    spy.mockClear();
  });

  const props = {
    authenticated: false,
    user: {},
    errors: {},
    fetchPopularBooks: jest.fn(),
    popularBooks: [],
    loadingPopularBooks: false
  }
  spy = jest.spyOn(HomePage.prototype, 'componentWillMount')
  const wrapper = shallow(<HomePage {...props} />)
  expect(HomePage.prototype.componentWillMount).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalled();
  expect(wrapper).toMatchSnapshot();
})

test('mapDispatchToProps', () => {
  const dispatch = jest.fn();
  expect(mapDispatchToProps(dispatch)).toHaveProperty('fetchPopularBooks');

  const { fetchPopularBooks } = mapDispatchToProps(dispatch);
  fetchPopularBooks();
  expect(dispatch).toHaveBeenCalled();
});

test('mapStateToProps', () => {
  expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('popularBooks');
  expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('loadingPopularBooks');
});

test('The home page component renders the top popular book card', () => {
  const props = {
    popularBooks: [{
      aboutAuthor: "Canadian journalist and writer.",
      author: "Malcolm Gladwell",
      borrowCount: 1,
      description: "Malcolm Gladwell redefined how we understand the world around us.",
      downVotes: 0,
      favorited: [],
      id: 3,
      image: "book.jpg",
      isbn: "7301294701859",
      publishedYear: "2005",
      quantity: 7,
      reviews: [],
      title: "Blink",
      upVotes: 0,
      votes: [],
    }],
    fetchPopularBooks: jest.fn(),
    authenticated: false,
    user: {},
    errors: {},
    loadingPopularBooks: false
  }
  const wrapper = shallow(<HomePage {...props} />);
  
  expect(wrapper.find(BookCard).length).toBe(1);
})

test('The home page component renders the loader while fetching top popular books ', () => {
  const props = {
    popularBooks: [],
    fetchPopularBooks: jest.fn(),
    authenticated: false,
    user: {},
    errors: {},
    loadingPopularBooks: true
  }
  const wrapper = shallow(<HomePage {...props} />);
  expect(wrapper.find(Loader).length).toBe(1);
})