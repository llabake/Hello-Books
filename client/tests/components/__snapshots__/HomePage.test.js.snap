// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage should call componentWillMount 1`] = `
<div>
  <Slider />
  <div>
    <h4
      className="center-align"
    >
      <span>
        Most Popular BOOKS
      </span>
    </h4>
  </div>
  <div
    className="divider"
  />
  <div
    className="section"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <h5
          className="center-align"
        >
           No books found 
        </h5>
      </div>
    </div>
  </div>
  <div
    className="divider"
  />
  <div
    className="section similar"
  >
    <h3
      className="center-align"
    >
      Based on readers favorites
    </h3>
    <div>
      <Connect(Carousel) />
    </div>
  </div>
</div>
`;
